Update "Point to Wordpress.com" button logic

gius80 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #96196 

## Proposed Changes

This PR implements a new behavior for `Point to WordPress.com` button for a specific domain.
With a single the button:
- restore default A and CNAME (www) records
- set WordPress.com nameservers

![tip-tool](https://github.com/user-attachments/assets/efa275d0-e5e0-4751-b290-57633c899056)

![dialog](https://github.com/user-attachments/assets/304f001e-5db8-47ca-a710-560ffe4669ba)

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

As detailed in #96196, we tried different approaches to reduce users friction, but the last implementation didn't yield the expected results.

## Testing Instructions
This PR depends on wpcom PR 174547-ghe-Automattic/wpcom/

To test this PR, you'd need a recently transferred domain that uses external nameservers with not standard A records and/or not standard www CNAME record.

If you don't have a transferred domain, you can force `transfer_status` property to `completed` in the backend.

Navigate site domain list (or all domains list) and verify that `Point to WordPress.com` button is showed in the domain row. Click on the button and confirm the operation on the dialog window.

Verify that all the changes have been done on domain settings page. Note that the button doesn't disappear immediately, since it requires dns propagation to reflect the changes in domains list.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

gius80 committed:
Update Point to Wordpress.com button logic

gius80 committed:
Call point-to-wpcom endpoint

gius80 committed:
Update dialog copy

gius80 committed:
Fix unit tests

gius80 commented:
Hey @fditrapani, here is the PR I created that should reduce friction for transferred domains. Feel free to leave any suggestion, in particular to UI and copy. Thanks in advance :)

fditrapani commented:
Thanks @gius80. I'll try take a look either later this week or next.

fditrapani commented:
One thing I noticed immediately was the modal and the old blue in the Confirm button. Can you update it to use the core modal component. We use that on on the plugins page, if you [manage your plugins](https://wordpress.com/plugins/manage/sites) and click on the activate button, it shows the right modal with the right colours:

<img width="1839" alt="image" src="https://github.com/user-attachments/assets/9c0804a0-6a91-4cd9-b19e-ebcd5350922c" />

gius80 committed:
Update code to use modal window

gius80 commented:
> One thing I noticed immediately was the modal and the old blue in the Confirm button. Can you update it to use the core modal component. We use that on on the plugins page, if you [manage your plugins](https://wordpress.com/plugins/manage/sites) and click on the activate button, it shows the right modal with the right colours:
Thanks @fditrapani! I've update the dialog to use the modal. I'm trying to figure it out what's the best way to update the color, since it's not specific to that single component but it depends on the color scheme you set on your user profile (see the screenshot after I choose the coffe color scheme):

<img width="597" alt="Screenshot 2025-02-27 alle 14 39 56" src="https://github.com/user-attachments/assets/d30f71f1-0447-4406-8188-49ef5326685f" />

leonardost commented on packages/domains-table/src/point-to-wpcom-dialog/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,48 @@
+					) }
+				</p>
+				<ul>
+					<li>Set WordPress.com defaults nameservers</li>
```
```suggestion
					<li>Set WordPress.com default nameservers</li>
```
Nit

leonardost replied:
```suggestion
					<li>{ __( 'Set WordPress.com default nameservers' ) }</li>
```
Also, these strings need to be translated

leonardost commented on packages/domains-table/src/point-to-wpcom-dialog/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,48 @@
+				</p>
+				<ul>
+					<li>Set WordPress.com defaults nameservers</li>
+					<li>Delete any existing A records</li>
```
```suggestion
					<li>{ __( 'Set A records to the default WordPress.com ones' ) }</li>
```
Suggestion

leonardost commented on packages/domains-table/src/point-to-wpcom-dialog/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,48 @@
+				<ul>
+					<li>Set WordPress.com defaults nameservers</li>
+					<li>Delete any existing A records</li>
+					<li>Delete any existing "www" CNAME record</li>
```
```suggestion
					<li>{ __( 'Set the default "www" CNAME record' ) }</li>
```
Suggestion

leonardost commented on packages/domains-table/src/point-to-wpcom-dialog/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,48 @@
+					<li>Delete any existing "www" CNAME record</li>
+				</ul>
+				<p>
+					{ __( 'These changes may take some time to apply. Are you sure you want to proceed?' ) }
```
```suggestion
					{ __( 'These changes may take some time to propagate. Are you sure you want to proceed?' ) }
```
Small suggestion

leonardost commented on packages/domains-table/src/domains-table/domains-table-row.tsx:
```diff
@@ -129,160 +136,191 @@ export function DomainsTableRow( { domain }: DomainsTableRowProps ) {
+					domain: domain.domain,
+				},
+			} );
+			dispatch( successNotice( __( 'Domain pointed to WordPress.com' ) ) );
```
```suggestion
			dispatch( successNotice( __( 'Domain is now pointing to WordPress.com' ) ) );
```
Maybe this reads more naturally?

leonardost reviewed (commented):
Tested and LGTM, thanks for working on this improvement! Left some comments in the code, mostly copy suggestions.

One weird thing is that the success notification didn't popup for me after I pressed the button 🤔  another thing is that, after you press the button, it stays there. I wonder if we should hide it or disable it for some time, otherwise maybe users will think it didn't work?

https://github.com/user-attachments/assets/1e701fda-6add-460b-ba85-0792d657ac8f

fditrapani commented:
Thanks for the update @gius80! Looks much better! Here's a suggestion for the copy:

**Title:** 
Point to WordPress.com

**Copy:** 
When you point your domain to WordPress.com, we will:
- Change your name servers to use the WordPress.com defaults,
- Delete any A records, and
- Delete any 'www' CNAME records.

Please note that these changes may take some time to apply. 

**Actions:** 
- Cancel
- Continue

> One weird thing is that the success notification didn't popup for me after I pressed the button 🤔 another thing is that, after you press the button, it stays there. I wonder if we should hide it or disable it for some time, otherwise maybe users will think it didn't work?

Good call out @leonardost, we should keep the button animating going until the action is complete and the confirmation notification is shown.

gius80 committed:
Fix notification

gius80 committed:
Update copy

gius80 committed:
Declare module

gius80 committed:
Declare module fix

gius80 committed:
Code refactoring

gius80 committed:
Clean type declaration

gius80 committed:
Update mobile card

gius80 committed:
Restore original empty lines

gius80 committed:
Add track event

gius80 commented:
Thanks for the review @fditrapani and @leonardost!

I updated the copy and fixed the success notification. To do that I had to completely refactor the code, since a I couldn't use the calypso redux state in the package, so I moved the logic outside.

@fditrapani regarding the button color, is it ok to keep the theme one for the moment? I guess this probably worths a wider discussion, what do you think?

gius80 commented:
> Thanks for those updates. This is a tricky one for me to test. Would you mind posting a quick video?

Sure! Here is the video @fditrapani:
https://d.pr/v/hQEoqi

> To answer your question about the colour, if it's coming from the modal component, then I wouldn't worry about changing it. These types of localized fixes cause us headaches down the line.

It's coming from the `is-primary` class associated to the button in the modal, which comes from the admin theme. At least, for all domains/purchases and other calypso pages. I can override the button color, but every other button in that page will be different.

fditrapani commented:
Thanks for sharing the video! Some really minor comments! 

1. I don't know if it's because of the recording but just wanted to confirm that the overlay covers the full screen and that the modal is centred? 
   
    <img width="1575" alt="image" src="https://github.com/user-attachments/assets/ae272342-65c9-4855-bded-d145a73ca8a9" />

2. After we show the success notification, can we hide the point button? With it still there, it looks like it didn't work.

> It's coming from the is-primary class associated to the button in the modal, which comes from the admin theme. At least, for all domains/purchases and other calypso pages. I can override the button color, but every other button in that page will be different.

No problem. Let's keep it as is for now.

gius80 commented:
> Thanks for sharing the video! Some really minor comments!
> 
> 1. I don't know if it's because of the recording but just wanted to confirm that the overlay covers the full screen and that the modal is centred?

Yes it's centered, I just cropped the left area from the video 😅

> 2. After we show the success notification, can we hide the point button? With it still there, it looks like it didn't work.
> 
> > It's coming from the is-primary class associated to the button in the modal, which comes from the admin theme. At least, for all domains/purchases and other calypso pages. I can override the button color, but every other button in that page will be different.
> 

This one is a bit tricky, since we show the button if the domain is not pointing to wpcom. It's a property that comes from the backend and relies on dns propagation, which can takes several minutes minutes. On the other hand, we can reset the cache value in the backend: this would hide the button, but the website will not resolving until dns propagation is completed. What do you think? Pinging @leonardost, maybe he has a better idea :D

gius80 committed:
Refetch domain after pointing to wpcom

leonardost commented on client/my-sites/domains/domain-management/components/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,42 @@
+				<p>{ __( 'When you point your domain to WordPress.com, we will:' ) }</p>
+				<ul>
+					<li>{ __( 'Change your name servers to use the WordPress.com defaults,' ) }</li>
+					<li>{ __( 'Delete any A records, and' ) }</li>
+					<li>{ __( "Delete any 'www' CNAME records" ) }</li>
```
When I read this, I get the feeling that the domain will end up with no `A` records and no `www` `CNAME` records 😅 what do you think of something like "Reset to default A records" and "Set the default 'www' CNAME record"?

gius80 replied:
Thant makes sense!

leonardost commented on client/my-sites/domains/domain-management/components/point-to-wpcom-dialog.tsx:
```diff
@@ -0,0 +1,42 @@
+};
+
+export const PointToWpcomDialog = ( { visible, onClose }: PointToWpcomDialogProps ) => {
+	const { __ } = useI18n();
```
```suggestion
	if ( ! visible ) {
		return null;
	}

	const { __ } = useI18n();
```
Suggestion: what do you think of adding an early return to avoid the ternary at the end of the component?

leonardost commented on packages/domains-table/src/domains-table/domains-table.tsx:
```diff
@@ -315,7 +321,6 @@ export const useGenerateDomainsTableState = ( props: DomainsTableProps ) => {
 	);
 
 	const [ updatingDomain, setUpdatingDomain ] = useState< Value[ 'updatingDomain' ] >( null );
-
```
Nit: can we add this newline again? 🙂

leonardost commented on client/my-sites/domains/domain-management/list/bulk-site-domains.tsx:
```diff
@@ -117,6 +125,26 @@ export default function BulkSiteDomains( props: BulkSiteDomainsProps ) {
+				},
+			} );
+			await refetch();
+			dispatch( successNotice( __( 'Domain pointed to WordPress.com' ) ) );
```
```suggestion
			dispatch( successNotice( __( 'Domain is pointing to WordPress.com' ) ) );
```
Suggestion

leonardost reviewed (approved):
Tested and LGTM, thanks for working on this! Left just minor suggestions and nitpicks.

fditrapani reviewed (approved):
> Thanks again [@fditrapani](https://github.com/fditrapani), here is a new video with the new implementation:

Amazing! looks great now on the front-end!

gius80 committed:
Address last review suggestions
